File: Top 100 Codes/03-Codes for Recursion/16-Find all possible Palindromic partitions of the given String/FindallpossiblePalindromicpartitionsofthegivenString.py
def is_palindrome(s, start, end):
    """Check if the substring s[start:end+1] is a palindrome."""
    while start < end:
        if s[start] != s[end]:
            return False
        start += 1
        end -= 1
    return True

def find_partitions(s, start, current, result):
    """Recursive function to find all palindromic partitions."""
    if start == len(s):
        result.append(current[:])  # Add a copy of the current partition to the result
        return
    
    for end in range(start, len(s)):
        if is_palindrome(s, start, end):
            current.append(s[start:end+1])  # Add the palindrome substring
            find_partitions(s, end + 1, current, result)  # Recur for the remaining string
            current.pop()  # Backtrack to try other partitions

def all_palindromic_partitions(s):
    """Function to print all palindromic partitions of a string."""
    result = []
    find_partitions(s, 0, [], result)
    
    print("All possible palindromic partitions:")
    for partition in result:
        print(" ".join(partition))

# Main function to take user input and call the partitioning function
if __name__ == "__main__":
    s = input("Enter a string: ")
    all_palindromic_partitions(s)

File: Top 100 Codes/03-Codes for Recursion/16-Find all possible Palindromic partitions of the given String/FindallpossiblePalindromicpartitionsofthegivenString.cpp
#include <bits/stdc++.h>
using namespace std;

// Function to check if a substring is a palindrome
bool isPalindrome(const string &str, int low, int high) {
    while (low < high) {
        if (str[low] != str[high]) 
            return false;
        low++;
        high--;
    }
    return true;
}

// Recursive function to find all palindromic partitions
void findAllPartitions(vector<vector<string>> &result, vector<string> &current, int start, int n, const string &str) {
    if (start >= n) {
        result.push_back(current);  // Store the current partition
        return;
    }

    for (int i = start; i < n; i++) {
        if (isPalindrome(str, start, i)) {
            current.push_back(str.substr(start, i - start + 1));  // Add palindrome substring
            findAllPartitions(result, current, i + 1, n, str);    // Recur for the remaining substring
            current.pop_back();  // Backtrack
        }
    }
}

// Main function to generate all palindromic partitions
void allPalPartitions(const string &str) {
    vector<vector<string>> result;
    vector<string> current;
    findAllPartitions(result, current, 0, str.length(), str);

    cout << "All possible palindromic partitions:\n";
    for (const auto &partition : result) {
        for (const string &s : partition) {
            cout << s << " ";
        }
        cout << "\n";
    }
}

int main() {
    string str;
    cout << "Enter a string: ";
    cin >> str;

    allPalPartitions(str);

    return 0;
}
